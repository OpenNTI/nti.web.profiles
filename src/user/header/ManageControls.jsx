import React from 'react';
import PropTypes from 'prop-types';
import {getService, User as UserResolver} from '@nti/web-client';
import {User, Layouts} from '@nti/web-commons';
import {scoped} from '@nti/lib-locale';

import MessageButton from './MessageButton';

const {Responsive} = Layouts;

const t = scoped('nti-web-profiles.user.header.ManageControls', {
	message: 'Message',
	follow: 'Follow',
	unfollow: 'Unfollow',
	isOffline: 'is offline',
	user: 'User'
});

export default class ManageControls extends React.Component {

	static propTypes = {
		entity: PropTypes.object.isRequired
	}

	state = {
		loading: true
	}

	componentDidMount () {
		const {entity} = this.props;

		this.setupFor(entity);
	}

	componentDidUpdate (prevProps) {
		if(prevProps.entity.getID() !== this.props.entity.getID()) {
			this.setupFor(this.props.entity);
		}
	}

	async setupFor (entity) {
		const service = await getService();
		const ds = service.getContacts();

		const resolved = await UserResolver.resolve({entity});

		if(this.unmounting) {
			return;
		}

		this.ds = ds;
		this.ds.addListener('change', this.onDataSourceChanged);
		this.isDesktop = !Responsive.isMobileContext();

		this.setState({loading: false, displayName: resolved.displayName, isContact: ds.contains(entity)});
	}

	componentWillUnmount () {
		this.unmounting = true;

		if(this.ds) {
			this.ds.removeListener('change', this.onDataSourceChanged);
		}
	}

	onDataSourceChanged = () => {
		this.setState({isContact: this.ds.contains(this.props.entity)});
	}

	follow = async () => {
		const {entity} = this.props;

		try {
			await this.ds.addContact(entity);
		}
		catch (e) {
			this.setState({error: e.message || 'Could not add contact'});
		}
	}

	unfollow = async () => {
		const {entity} = this.props;

		try {
			await this.ds.removeContact(entity);
		}
		catch (e) {
			this.setState({error: e.message || 'Could not remove contact'});
		}
	}

	renderNonContactButtons () {
		return (
			<div className="non-contact profile-manage-controls">
				<div className="nti-button follow" onClick={this.follow}><i className="icon-addfriend"/>{t('follow')}</div>
			</div>
		);
	}

	renderContactButtons () {
		const {entity} = this.props;
		const {displayName} = this.state;

		return (
			<div className="contact profile-manage-controls">
				{this.isDesktop && (
					<User.Presence user={entity}>
						<MessageButton entity={entity} displayName={displayName || t('user')}/>
					</User.Presence>
				)}
				<div className="nti-button unfollow" onClick={this.unfollow}><i className="icon-unfollow"/>{t('unfollow')}</div>
			</div>
		);
	}

	render () {
		const {loading, isContact} = this.state;

		if(loading) {
			return null;
		}

		return isContact ? this.renderContactButtons() : this.renderNonContactButtons();
	}
}
