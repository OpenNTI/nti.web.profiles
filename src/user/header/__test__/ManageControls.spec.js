import React from 'react';
import renderer from 'react-test-renderer';
import { TestUtils } from '@nti/web-client';

import ManageControls from '../ManageControls';

const { tearDownTestClient, setupTestClient } = TestUtils;

const getMockService = isContact => {
	return {
		getContacts: () => {
			return {
				addListener: () => {},
				removeListener: () => {},
				contains: () => isContact,
			};
		},
	};
};

const onBefore = isContact => {
	jest.useFakeTimers();
	setupTestClient(getMockService(isContact));
};

const onAfter = () => {
	tearDownTestClient();
};

const flushPromises = () => new Promise(resolve => setImmediate(resolve));

/* eslint-env jest */
describe('User profile manage controls test (is not a contact)', () => {
	beforeEach(() => onBefore(false));
	afterEach(onAfter);

	test('Basic render test', async () => {
		const user = {
			Username: 'testUser',
			getID: () => 'testUser',
		};

		const cmp = renderer.create(<ManageControls entity={user} />);

		jest.runAllTimers();
		await flushPromises();
		jest.runAllTimers();

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});
});

describe('User profile manage controls test (is a contact)', () => {
	beforeEach(() => onBefore(true));
	afterEach(onAfter);

	test('Basic render test', async () => {
		const user = {
			Username: 'testUser',
			getID: () => 'testUser',
		};

		const cmp = renderer.create(<ManageControls entity={user} />);

		jest.runAllTimers();
		await flushPromises();
		jest.runAllTimers();

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
