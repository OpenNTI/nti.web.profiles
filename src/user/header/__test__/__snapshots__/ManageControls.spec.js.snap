// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User profile manage controls test (is a contact) Basic render test 1`] = `
<div
  className="contact profile-manage-controls"
>
  <a
    className="nti-button flyout-trigger message message-button-flyout flyout-closed button primary"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="button"
    tabIndex={0}
  >
    <span>
      <div
        className="icon"
      />
      <span
        className="text"
      >
        Message
      </span>
    </span>
  </a>
  <a
    className="nti-button unfollow button primary"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="content"
    >
      <i
        className="icon-unfollow"
      />
      <span>
        Unfollow
      </span>
    </div>
  </a>
</div>
`;

exports[`User profile manage controls test (is not a contact) Basic render test 1`] = `
<div
  className="non-contact profile-manage-controls"
>
  <a
    className="nti-button follow button primary"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="content"
    >
      <i
        className="icon-addfriend"
      />
      <span>
        Follow
      </span>
    </div>
  </a>
</div>
`;
